File: sdks/python/src/etrace/instrumentation/anthropic.py
"""Anthropic auto-instrumentor.

Patches Anthropic messages (sync + async).
Uses etrace spans directly (no OTel dependency).
"""

from __future__ import annotations

import json
import logging
from typing import Any, ClassVar

from .base import BaseInstrumentor

logger = logging.getLogger("etrace.instrumentation")


class AnthropicInstrumentor(BaseInstrumentor):
    """Patches Anthropic messages (sync + async)."""

    name: ClassVar[str] = "anthropic"
    target_packages: ClassVar[list[str]] = ["anthropic"]

    def instrument(self, calc_costs: bool = True) -> bool:
        try:
            from anthropic.resources.messages import AsyncMessages, Messages
        except ImportError:
            logger.debug("anthropic not installed — skipping instrumentation")
            return False

        try:
            self._patch(
                Messages,
                "create",
                self._create_llm_wrapper_factory(
                    "anthropic.messages",
                    "anthropic",
                    self._process_messages,
                    calc_costs=calc_costs,
                    is_async=False,
                ),
            )
            self._patch(
                AsyncMessages,
                "create",
                self._create_llm_wrapper_factory(
                    "anthropic.messages",
                    "anthropic",
                    self._process_messages,
                    calc_costs=calc_costs,
                    is_async=True,
                ),
            )
        except Exception as exc:
            logger.warning("anthropic patch failed: %s", exc)
            self.uninstrument()
            return False

        logger.info("anthropic auto-instrumented (messages)")
        return True

    def _process_messages(
        self,
        span: Any,
        result: Any,
        kwargs: dict[str, Any],
        calc_costs: bool,
    ) -> None:
        model = self._resolve_model(result, str(kwargs.get("model", "")))

        if kwargs.get("stream", False):
            span.attributes["gen_ai.streaming"] = True
            return

        usage = getattr(result, "usage", None)
        if usage:
            input_tokens = getattr(usage, "input_tokens", 0) or 0
            output_tokens = getattr(usage, "output_tokens", 0) or 0
            cache_read = getattr(usage, "cache_read_input_tokens", 0) or 0
            cache_creation = getattr(usage, "cache_creation_input_tokens", 0) or 0

            self._set_usage_and_cost(
                span,
                model,
                input_tokens,
                output_tokens,
                cached_tokens=cache_read,
                cache_write_tokens=cache_creation,
                calc_costs=calc_costs,
            )

        try:
            content = getattr(result, "content", None)
            if content:
                text = "".join(block.text for block in content if hasattr(block, "text") and block.text)
                if text:
                    self._capture_output(span, text)
                else:
                    # When no text, the LLM is requesting tool use.
                    tool_blocks = [b for b in content if getattr(b, "type", None) == "tool_use"]
                    if tool_blocks:
                        tc_data = [
                            {"name": getattr(b, "name", ""), "input": getattr(b, "input", {})} for b in tool_blocks
                        ]
                        self._capture_output(span, json.dumps(tc_data))
        except Exception:
            pass

File: sdks/typescript/src/instrumentation/anthropic.ts
/**
 * Anthropic auto-instrumentor (TypeScript).
 *
 * Patches:
 *   - Anthropic.Messages.prototype.create
 *
 * Uses etrace spans (no OTel dependency).
 */
import type { Span } from "../types.js";
import { BaseInstrumentor } from "./base.js";

export class AnthropicInstrumentor extends BaseInstrumentor {
  name = "anthropic";
  targetPackages = ["@anthropic-ai/sdk"];

  instrument(calcCosts: boolean): boolean {
    const anthropic = this._resolve("@anthropic-ai/sdk");
    if (!anthropic) return false;

    const SDK = (anthropic.default ?? anthropic) as Record<string, unknown>;

    try {
      // ── Messages ──────────────────────────────────────────────────────
      const Messages = SDK.Messages as Record<string, unknown> | undefined;
      if (Messages?.prototype) {
        this.patch(Messages.prototype as Record<string, unknown>, "create", (original) =>
          this.wrapLLMCall(
            calcCosts,
            original,
            "anthropic.messages",
            "anthropic",
            (span, res, model) => {
              this.processMessages(span, res, model, calcCosts);
            },
          ),
        );
      }
    } catch (exc) {
      console.warn(`[etrace] anthropic patch failed: ${exc}`);
      this.uninstrument();
      return false;
    }

    console.info("[etrace] anthropic auto-instrumented (messages)");
    return true;
  }

  // ── Messages response processing ────────────────────────────────────────

  private processMessages(
    span: Span,
    result: unknown,
    requestModel: string,
    calcCosts: boolean,
  ): void {
    const res = result as Record<string, unknown> | null | undefined;
    if (!res) return;

    const model = (res.model as string) ?? requestModel;

    const isStream =
      typeof (res as unknown as AsyncIterable<unknown>)[Symbol.asyncIterator] === "function";
    if (isStream) {
      span.attributes ??= {};
      span.attributes["gen_ai.streaming"] = true;
      return;
    }

    // Extract usage
    const usage = res.usage as Record<string, unknown> | null | undefined;
    if (usage) {
      const inputTokens = (usage.input_tokens as number) ?? 0;
      const outputTokens = (usage.output_tokens as number) ?? 0;
      const cacheRead = (usage.cache_read_input_tokens as number) ?? 0;
      const cacheCreation = (usage.cache_creation_input_tokens as number) ?? 0;

      this.setUsageAndCost(
        span,
        model,
        inputTokens,
        outputTokens,
        0,
        cacheRead,
        0,
        cacheCreation,
        calcCosts,
      );
    }

    // Capture output text
    try {
      const content = res.content as Array<Record<string, unknown>> | null;
      if (content) {
        const text = content
          .filter((block) => typeof block.text === "string")
          .map((block) => block.text as string)
          .join("");
        if (text) {
          this.captureOutput(span, text);
        }
      }
    } catch {
      /* best-effort */
    }
  }
}
